### PR TITLE
OpenCL GPU class python wrapping

### PR DESCRIPTION
**Pull request by thewtex**

### CODE REVIEW

**Comment by dzenanz** [comment]
No Azure pipeline builds? Strange.

And this PR needs a rebase to avoid reviewing already merged commits.

**Comment by dzenanz** [comment]
The force-push did a rebase on current master. This also triggered AZP builds.

File: Modules/Core/GPUFiniteDifference/src/CMakeLists.txt [hunk 1]
```diff
@@ -7,10 +7,7 @@ if (ITK_USE_GPU)
     GPUDenseFiniteDifferenceImageFilter.cl
     )
   write_gpu_kernels("${ITKGPUFiniteDifference_Kernels}" ITKGPUFiniteDifference_SRCS)
-endif()
-
-itk_module_add_library(ITKGPUFiniteDifference ${ITKGPUFiniteDifference_SRCS})
```
[the remark this answers is not in the document]
Comment (thewtex) [reply Modules/Core/GPUFiniteDifference/src/CMakeLists.txt]: Nope -- this is never reached if `ITK_USE_GPU` is `OFF`. In general, we need all the components to build the library to build it successfully. I think this issue was discovered when building the modules externally.

File: Modules/Core/GPUCommon/CMakeLists.txt [hunk 2]
```diff
@@ -5,6 +5,7 @@ if(NOT ITK_SOURCE_DIR)
   find_package(ITK REQUIRED)
   list(APPEND CMAKE_MODULE_PATH ${ITK_CMAKE_DIR})
   set(ITK_USE_GPU ON CACHE BOOL "Enable OpenCL GPU support." FORCE)
+  set(ITK_USE_GPU ON)
```
[the remark this answers is not in the document]
Comment (thewtex) [reply Modules/Core/GPUCommon/CMakeLists.txt]: Yes -- cache and non-cache versions of the variable.

**Comment by thewtex** [comment]
@dzenanz thanks for the review and branch update :+1:

File: Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx [hunk 3]
```diff
@@ -324,8 +324,79 @@ GPUKernelManager::SetKernelArg(int kernelIdx, cl_uint argIdx, size_t argSize, co
   return true;
 }
 
+
+bool
+GPUKernelManager::SetKernelArgWithChar(int kernelIdx, cl_uint argIdx, char argVal)
+{
+  return this->SetKernelArg(kernelIdx, argIdx, sizeof(char), &argVal);
```
Comment (N-Dekker) [inline Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx]: Is it really necessary to pass the size of `argVal`  each time `this->SetKernelArg` is called (10 times)? It looks to me like a template function could be of help here.

File: Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx [hunk 4]
```diff
@@ -324,8 +324,79 @@ GPUKernelManager::SetKernelArg(int kernelIdx, cl_uint argIdx, size_t argSize, co
...
+GPUKernelManager::SetKernelArgWithFloat(int kernelIdx, cl_uint argIdx, float argVal)
+{
+  return this->SetKernelArg(kernelIdx, argIdx, sizeof(float), &argVal);
+}
+
+
+bool
+GPUKernelManager::SetKernelArgWithDouble(int kernelIdx, cl_uint argIdx, double argVal)
+{
+  return this->SetKernelArg(kernelIdx, argIdx, sizeof(double), &argVal);
```
Comment (N-Dekker) [inline Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx]: Please see my previous comment: passing the size of argVal each time seems redundant to me.

File: Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx (hunk 3 above)
Comment (dzenanz) [reply Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx]: For C++ it is totally redundant, I though that commit message said it eases access from Python.

Comment (N-Dekker) [reply Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx]: Sorry I don't have time for a complete review of all these (1000+) lines of code, so I just focused on some of the C++ work in there.  These ten `sizeof()`'s appear redundant and avoidable. As far as I can see, these are part of the C++ implementation, not the Python interface.

Comment (dzenanz) [reply Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx]: Oh, now I see what you meant.

Comment (thewtex) [reply Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx]: @N-Dekker addressed in 7ba119454b78dddd9d9f5d549207ba4030e09753

File: Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx (hunk 4 above)
Comment (thewtex) [reply Modules/Core/GPUCommon/src/itkGPUKernelManager.cxx]: Addressed in 7f0bf669b578b0b1ea315cd12a0ebeedf4d5b6bb

**Pull request merged**